### PR TITLE
Get derivatives of lapse from GH variables

### PR DESCRIPTION
**Pull request by prayush**
## Proposed changes

- Adds functions that compute time/spatial derivatives of the lapse function from generalized harmonic variables.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp [hunk 1]
```diff
@@ -166,6 +166,49 @@ tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
...
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+Scalar<DataType> time_deriv_of_lapse(
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  auto dt_lapse = make_with_value<Scalar<DataType>>(get(lapse), 0.);
+  for (size_t a = 0; a < SpatialDim+1; ++a) {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp]: spaces around the `+` (we have a clang_format file in the repo to help with these things :) )

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp [hunk 2]
```diff
@@ -166,6 +166,49 @@ tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
...
 
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::i<DataType, SpatialDim, Frame> spatial_deriv_of_lapse(
+    const Scalar<DataType>& lapse,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
+  auto deriv_lapse =
+    make_with_value<tnsr::i<DataType, SpatialDim, Frame>>(get(lapse), 0.0);
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t a = 0; a < SpatialDim+1; ++a) {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp]: spaces around the `+` (we have a clang_format file in the repo to help with these things :) )

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp [hunk 3]
```diff
@@ -166,6 +166,49 @@ tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
...
+      for (size_t b = 0; b < SpatialDim+1; ++b) {
+        deriv_lapse.get(i) += phi.get(i, a, b)
+          * spacetime_unit_normal.get(a) * spacetime_unit_normal.get(b);
+      }
+    }
+    deriv_lapse.get(i) *= (-0.5 * get(lapse));
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp]: More of an FYI, you can omit the outermost parens if you want: `*= -0.5 * get(lapse);` should work

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp [hunk 4]
```diff
@@ -102,6 +102,31 @@ void test_compute_extrinsic_curvature_and_deriv_metric(const T& used_for_size) {
   CHECK_ITERABLE_APPROX(deriv_spatial_metric, deriv_spatial_metric_test);
 }
 
+
+template <typename DataType, size_t SpatialDim, typename Frame>
+void test_3p1_deriv_functions(const DataVector& used_for_size) noexcept {
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp]: maybe `3plus1` to make it clearer for people :)

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 5]
```diff
@@ -129,4 +129,62 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
...
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept;
+
+/*!
+ * \ingroup GeneralRelativityGroup
+ * \brief Computes time derivative of lapse (N) from the generalized
+ *        harmonic variables, lapse, shift and the spacetime unit normal 1-form.
+ *
+ * \details Let the generalized harmonic conjugate momentum and spatial
+ * derivative variables be \f$\pi_{ab} = -n^c \partial_c \psi_{ab} \f$ and
+ * \f$\phi_{iab} = \partial_i \psi_{ab} \f$, and the operator
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: We've been using capital Pi and Phi

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 6]
```diff
@@ -129,4 +129,62 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
...
+ * \f$D := \partial_0 - N^k \partial_k \f$. The time derivative of N is then:
+ * \f{align*}
+ *  N^2/2 n^a n^b \pi_{ab} - N N^i/2 n^a n^b \phi_{iab}
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: `n^a` with `t^a` (we use `n` as the spatial normal to an element) and all the SpEC papers use `t^a`

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 7]
```diff
@@ -129,4 +129,62 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
...
+ *  =& N^2/2 n^a n^b n^c \partial_c \psi_{ab} - N N^i/2 (-(2/N) \partial_i N) \\
+ *  =& N^2/2 [-1/N^3 D[g_{jk} N^j N^k - N^2] - (N^j N^k / N^3)D[g_{jk}]
+ *           + 2(N^j / N^3) D[g_{jk} N^k] + (2 / N^2)(N^i \partial_i N)] \\
+ *  =& 1/(2N) [-D[g_{jk}N^jN^k - N^2] - N^jN^k D[g_{jk}]
+ *            + 2N N^k\partial_k N + 2N^j D[g_{jk}N^k]] \\
+ *  =& D[N] + N^k\partial_k N \\
+ *  =& \partial_0 N
+ * \f}
+ * where the simplification done for \f$\partial_i N\f$ is used to substitute
+ * for the second term (\f$N N^i/2 n^a n^b \phi_{iab}\f$).
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]:
```
Thus,
\f[
\partial_t N = (N/2)(N t^a t^b \Pi_{ab} - N^i t^a t^b \Phi_{iab})
\f]
```

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 8]
```diff
@@ -129,4 +129,62 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
 template <size_t SpatialDim, typename Frame, typename DataType>
 tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
     const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept;
+
+/*!
+ * \ingroup GeneralRelativityGroup
+ * \brief Computes spatial derivatives of lapse (N) from the generalized
+ *        harmonic variables and spacetime unit normal 1-form.
+ *
+ * \details If the generalized harmonic conjugate momentum and spatial
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:138]: Same comments about as for the time deriv one below :)

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp [hunk 9]
```diff
@@ -102,6 +102,31 @@ void test_compute_extrinsic_curvature_and_deriv_metric(const T& used_for_size) {
...
+  pypp::check_with_random_values<1>(
+      static_cast<Scalar<DataType> (*)(
+        const Scalar<DataType>&,
+        const tnsr::I<DataType, SpatialDim, Frame>&,
+        const tnsr::A<DataType, SpatialDim, Frame>&,
+        const tnsr::iaa<DataType, SpatialDim, Frame>&,
+        const tnsr::aa<DataType, SpatialDim, Frame>&)>(
+      &::GeneralizedHarmonic::time_deriv_of_lapse<SpatialDim, Frame, DataType>),
+      "TestFunctions", "dt_lapse",
+      {{{std::numeric_limits<double>::denorm_min(), 10.0}}}, used_for_size);
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp]: Could you verify that you get the right answer for Kerr-Schild?

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp [hunk 10]
```diff
@@ -166,6 +166,49 @@ tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
...
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  auto dt_lapse = make_with_value<Scalar<DataType>>(get(lapse), 0.);
+  for (size_t a = 0; a < SpatialDim+1; ++a) {
+    for (size_t b = 0; b < SpatialDim+1; ++b) {
+      // first term
+      get(dt_lapse) += 0.5 * get(lapse) * get(lapse) * pi.get(a, b)
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp]: `square(get(lapse))`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp [hunk 11]
```diff
@@ -166,6 +166,49 @@ tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
...
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+Scalar<DataType> time_deriv_of_lapse(
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  auto dt_lapse = make_with_value<Scalar<DataType>>(get(lapse), 0.);
```
Comment (fmahebert) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:195]: Style nitpick: we allow either `0.0` or `0.`, but the choice should be made consistently throughout the file

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp [hunk 12]
```diff
@@ -166,6 +166,49 @@ tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
   return deriv_spatial_metric;
 }
 
+template <size_t SpatialDim, typename Frame, typename DataType>
```
Comment (fmahebert) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:169]: What is the reasoning behind this ordering of the template parameters? It might be nice to keep it consistent with the tensor's parameters, i.e. `<DataType, SpatialDim, Frame>`.

Comment (fmahebert) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:169]: Oh, I guess this is how the rest of the file orders these.. :-/

Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:169]: yeah, not just this file but all code in evolution system GeneralizedHarmonic follows this ordering, which is the precedent i followed

Comment (fmahebert) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:169]: Definitely the right choice for this PR then 👍

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp (hunk 10 above)
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp]: either that, or the `0.5 * get(lapse)` can be factored out of both loops and just multiplied at the end.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 13]
```diff
@@ -129,4 +129,62 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
...
+ * derivative variables are \f$\pi_{ab} = -n^c \partial_c \psi_{ab} \f$ and
+ * \f$\phi_{iab} = \partial_i \psi_{ab} \f$, the spatial derivatives of N
+ * can be obtained from:
+ * \f{align*}
+ *  n^a n^b \phi_{iab} = -1/(2N) [\partial_i (-N^2 + N_jN^j)-
+ *                               2 N^j \partial_i N_j
+ *                               + N^j N^k \partial_i g_{jk}]
+ *                     = -(2/N) \partial_i N
+ * \f},
```
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: comma should be on line with equation, `since` on its own line, and the final equation on `\f[` `\f]`.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 14]
```diff
@@ -129,4 +129,62 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
...
+ * since
+ * \f$\partial_i (N_jN^j) = 2N^j \partial_i N_j - N^j N^k \partial_i g_{jk} \f$.
+ *
+ * ==> \f$ \partial_i N = -(N/2) n^a \phi_{iab} n^b \f$
```
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: ==> can be `\Longrightarrow` in the equation.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 6 above)
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: the `N^2/2` that appear in several places would be more readable if you used instead `\frac{N^2}{2}` or `\frac{1}{2} N^2`.

**Review by moxcodes — COMMENTED** [review]
Group code review with @markscheel and Rob.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 6 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: changed to `(1/2) N^2` to match rest of the doc text.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 15]
```diff
@@ -129,4 +129,68 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
...
+ *                               2 N^j \partial_i N_j
+ *                               + N^j N^k \partial_i g_{jk}]
+ *                     = -(2/N) \partial_i N,
+ * \f}
+ * since
+ * \f[
+ * \partial_i (N_jN^j) = 2N^j \partial_i N_j - N^j N^k \partial_i g_{jk}.
+ * \f]
+ *
+ * \f$ \Longrightarrow \partial_i N = -(N/2) t^a \Phi_{iab} t^b \f$
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: Could you make this `\f[ ... \f]` so it's centered in the docs? :)

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 16]
```diff
@@ -129,4 +129,68 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
...
+ *        harmonic variables, lapse, shift and the spacetime unit normal 1-form.
+ *
+ * \details Let the generalized harmonic conjugate momentum and spatial
+ * derivative variables be \f$\Pi_{ab} = -t^c \partial_c \psi_{ab} \f$ and
+ * \f$\Phi_{iab} = \partial_i \psi_{ab} \f$, and the operator
+ * \f$D := \partial_0 - N^k \partial_k \f$. The time derivative of N is then:
+ * \f{align*}
+ *  (1/2) N^2 t^a t^b \Pi_{ab} - N N^i/2 t^a t^b \Phi_{iab}
+ *  =& (1/2) N^2 t^a t^b t^c \partial_c \psi_{ab}
+ *       - N N^i/2 (-(2/N) \partial_i N) \\
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: I find `N N^i/2 (...)` confusing. Are you dividing by everything after the `/` or just the `2`? I think doing as @moxcodes suggested and using `\frac` (even if it's for now inconsistent with the rest of the file) would make the equations easier to parse :) I appreciate you actually showing the path to the result and not just the result!

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp [hunk 17]
```diff
@@ -102,6 +103,31 @@ void test_compute_extrinsic_curvature_and_deriv_metric(const T& used_for_size) {
   CHECK_ITERABLE_APPROX(deriv_spatial_metric, deriv_spatial_metric_test);
 }
 
+
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp]: Only one space between functions :)

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp [hunk 18]
```diff
@@ -102,6 +103,31 @@ void test_compute_extrinsic_curvature_and_deriv_metric(const T& used_for_size) {
...
+        const Scalar<DataType>&,
+        const tnsr::I<DataType, SpatialDim, Frame>&,
+        const tnsr::A<DataType, SpatialDim, Frame>&,
+        const tnsr::iaa<DataType, SpatialDim, Frame>&,
+        const tnsr::aa<DataType, SpatialDim, Frame>&)>(
+      &::GeneralizedHarmonic::time_deriv_of_lapse<SpatialDim, Frame, DataType>),
+      "TestFunctions", "dt_lapse",
+      {{{std::numeric_limits<double>::denorm_min(), 10.}}}, used_for_size);
+}
+
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp:143]: Still need the Kerr-Schild test :)

**Review by nilsdeppe — COMMENTED** [review]
I'm happy with the changes. A couple minor stylistic points for the dox. The biggest thing I'd still like to see is a Kerr-Schild test. We've been burned a lot of times now with just the python tests. I believe the [documentation for the Kerr-Schild analytic solution](https://spectre-code.org/classgr_1_1solutions_1_1kerrschild) has the derivatives worked out :)

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp [hunk 19]
```diff
@@ -99,7 +99,7 @@ tnsr::a<DataType, SpatialDim, Frame> gauge_source(
         trace_christoffel_last_indices) noexcept {
   DataType one_over_lapse = 1.0 / get(lapse);
   auto gauge_source_h =
-      make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(lapse, 0.0);
+      make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(lapse, 0.);
```
Comment (geoffrey4444) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:102]: Why this change?

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 20]
```diff
@@ -129,4 +129,68 @@ tnsr::ii<DataType, SpatialDim, Frame> extrinsic_curvature(
 template <size_t SpatialDim, typename Frame, typename DataType>
 tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
     const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept;
+
+/*!
+ * \ingroup GeneralRelativityGroup
+ * \brief Computes spatial derivatives of lapse (N) from the generalized
```
Comment (geoffrey4444) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:135]: I prefer `\f$alpha\f$` and `\f$beta\f$` for lapse and shift, but I’ll defer if everyone else disagrees.

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp (hunk 19 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:102]: I'm guessing this is to make the `0.0` and `0.` consistent in the file.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 20 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:135]: Probably a philosophical debate, but we currently use both `N` and `alpha` in different parts of the code, so we should file an issue to make the code base consistent (whatever is decided on). Does that sound okay with you, @geoffrey4444 ?

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp (hunk 19 above)
Comment (fmahebert) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:102]: Unfortunately the file wasn't actually consistent beforehand (though primarily used `0.0`)... and still isn't now.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 20 above)
Comment (geoffrey4444) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:135]: Yep!

**Comment by nilsdeppe** [comment]
@prayush I'm marking this as "in progress" until the Kerr-Schild test exists. If you are unable to remove the label when you push that test please just ping me and I'll do it. I'm going to do the same for the other PRs that add space and time derivs of the 3+1 quantities since they'll all need basically the same changes :)

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp (hunk 9 above)
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp]: Just added KerrSchild test.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 15 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: Changed

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 16 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: Agreed, changed to `\frac` .

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp (hunk 17 above)
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp]: fixed

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp (hunk 18 above)
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp:143]: Just added it.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp (hunk 19 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:102]: It should be consistent now with the use of `0.` instead of `0.0`.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 20 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:135]: I actually prefer `N` and `\vec{N}` for the two, but this is a question for broader consensus..

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 8 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:138]: Just added KerrSchild test.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp (hunk 11 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp:195]: I've made it consistently `0.`

**Comment by prayush** [comment]
> @prayush I'm marking this as "in progress" until the Kerr-Schild test exists. If you are unable to remove the label when you push that test please just ping me and I'll do it. I'm going to do the same for the other PRs that add space and time derivs of the 3+1 quantities since they'll all need basically the same changes :)

@nilsdeppe - thanks for doing that. I've just pushed Kerr-Schild tests for this and other PRS that add space/time derivs of 3+1 quantities. Could you please remove the "In progress" flag? Also, is it possible to grant me the necessary permissions to add/subtract these flags to (self owned) PRs?

**Comment by nilsdeppe** [comment]
> > @prayush I'm marking this as "in progress" until the Kerr-Schild test exists. If you are unable to remove the label when you push that test please just ping me and I'll do it. I'm going to do the same for the other PRs that add space and time derivs of the 3+1 quantities since they'll all need basically the same changes :)
> 
> @nilsdeppe - thanks for doing that. I've just pushed Kerr-Schild tests for this and other PRS that add space/time derivs of 3+1 quantities. Could you please remove the "In progress" flag? Also, is it possible to grant me the necessary permissions to add/subtract these flags to (self owned) PRs?

Okay, you should have the power after you accept the team invitation I sent :)

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp [hunk 21]
```diff
@@ -168,18 +170,19 @@ tnsr::i<DataType, SpatialDim, Frame> spatial_deriv_of_lapse(
...
- *  =& (1/2) N^2 t^a t^b t^c \partial_c \psi_{ab}
- *       - N N^i/2 (-(2/N) \partial_i N) \\
- *  =& (1/2) N^2 [-1/N^3 D[g_{jk} N^j N^k - N^2] - (N^j N^k / N^3)D[g_{jk}]
- *           + 2(N^j / N^3) D[g_{jk} N^k] + (2 / N^2)(N^i \partial_i N)] \\
- *  =& 1/(2N) [-D[g_{jk}N^jN^k - N^2] - N^jN^k D[g_{jk}]
+ *  \frac{1}{2} N^2 t^a t^b \Pi_{ab} - \frac{1}{2} N N^i t^a t^b \Phi_{iab}
+ *  =& \frac{1}{2} N^2 t^a t^b t^c \partial_c \psi_{ab}
+ *       - \frac{1}{2} N N^i (-(2/N) \partial_i N) \\
+ *  =& \frac{1}{2} N^2 [-(1/N^3) D[g_{jk} N^j N^k - N^2] \\
+ *           - (N^j N^k / N^3)D[g_{jk}] \\
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: I think you're missing a `&` before the `-`. i.e.
```cpp
 *           &- (N^j N^k / N^3)D[g_{jk}] \\
```
same with the next line. Just a small typo :)

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp [hunk 22]
```diff
@@ -128,6 +141,68 @@ void test_lapse_deriv_functions(const DataVector& used_for_size) noexcept {
...
 }
 
+// Test computation of derivs of lapse by comparing to Kerr-Schild
+template <typename Solution>
+void test_lapse_deriv_functions_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound) noexcept {
+  // Setup grid
+  const size_t SpatialDim = 3;
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp]: `spatial_dim` (we use snake_case for variables and CamelCase for all template parameters, just so you know why it's CamelCase in a lot of places :) )

**Review by nilsdeppe — COMMENTED** [review]
A couple very minor comments. I'm happy for you to squash everything into one commit once @geoffrey4444 is happy with your changes :)

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 21 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp]: fixed

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp (hunk 22 above)
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp]: oops, fixed now

**Comment by nilsdeppe** [comment]
@geoffrey4444 could you please check if you're happy with these changes when you have time? Thanks :)

**Comment by kidder** [comment]
I am also happy for you to squash when @geoffrey4444 gives you the go-ahead

**Review by geoffrey4444 — DISMISSED** [review]

**Comment by kidder** [comment]
@prayush please squash this into a single commit.

**Comment by prayush** [comment]
> @prayush please squash this into a single commit.

@kidder  - Squashed into one commit.

File: src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp (hunk 20 above)
Comment (kidder) [reply src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:135]: Issue opened: #1209

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@nilsdeppe this is ready

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
restarted gcc-6 release build.

**Review by wthrowe — APPROVED** [review]
